### PR TITLE
fix: not found in DEV and refactor fallback handling

### PR DESCRIPTION
**Pull request by dai-shi**
following up #809.

This is still incomplete and need more refactors as noted in `FIXME`.

### CODE REVIEW

File: packages/waku/src/cli.ts [hunk 1]
```diff
@@ -96,9 +96,14 @@ async function runDev() {
...
-  // @ts-expect-error FIXME there might be a better way to handle this
-  app.use('*', (_c, _next) => {});
+  app.notFound((c) => {
+    // FIXME can we avoid hardcoding the public path?
+    const file = path.join('public', '404.html');
+    if (existsSync(file)) {
+      return c.html(readFileSync(file, 'utf8'), 404);
+    }
+    return c.text('404 Not Found', 404);
+  });
```
Comment (dai-shi) [inline packages/waku/src/cli.ts:106]: I wonder if this breaks what we fixed in #796.

File: packages/waku/src/lib/middleware/dev-server-impl.ts [hunk 2]
```diff
@@ -369,6 +368,7 @@ export const devServer: Middleware = (options) => {
 
     if (
       // HACK depending on `rscPath` is a bad idea
+      // FIXME This hack should be removed as well as `willBeHandled`
       ctx.req.url.pathname.startsWith(config.basePath + config.rscPath + '/') ||
       !(await willBeHandled(ctx.req.url.pathname))
```
Comment (dai-shi) [inline packages/waku/src/lib/middleware/dev-server-impl.ts:373]: I was hoping to remove this, but I'm not sure how to deal with it with Vite. So, I hope someone will help in the future.

**Comment by dai-shi** [comment]
I hope this doesn't break #644 and others, but not 100% sure. cc @rmarscher 
Let's merge and see if we get any feedback.

**Comment by rmarscher** [comment]
I deployed a test with 0.21.0-beta.5 and it still responds with the expected 404 error and not a 200. 👍

**Pull request merged**